### PR TITLE
Add Django 3.1 support

### PR DESCRIPTION
**Pull request by FinalAngel**

### CODE REVIEW

**Comment by yakky** [comment]
@FinalAngel LOL I just opened #1202  for 3.1 :D

**Comment by yakky** [comment]
@FinalAngel I switched #1202 as a PR against your branch

**Comment by FinalAngel** [comment]
@yakky lol I just pushed a bunch of stuff too :D

**Comment by FinalAngel** [comment]
@yakky checkout this small beast:
https://github.com/divio/django-filer/pull/1201/commits/39621f5fbb3e5ba364f2627acb81dc390171b92b

:D

**Comment by yakky** [comment]
@FinalAngel I closed mine :)

File: .travis.yml [hunk 1]
```diff
@@ -1,41 +1,44 @@
...
-    - python: 3.6
+    - python: 3.5
       env: TOX_ENV='frontend'
-    # Django 1.11
-    - python: 3.4
-      env: DJANGO='dj111' SWAP='noswap'
+    # Django 2.2
     - python: 3.5
-      env: DJANGO='dj111' SWAP='noswap'
+      env: DJANGO='dj22' CMS='cms37' SWAP='swap'
```
Comment (yakky) [inline .travis.yml:17]: filer does not depend on django CMS, do we have to introduce it in the build matrix?

File: filer/admin/patched/admin_utils.py [hunk 2]
```diff
@@ -106,12 +98,12 @@ def collect(self, objs, source=None, source_attr=None, **kwargs):
...
             self.model_objs[obj._meta.model].add(obj)
         try:
-            return super(NestedObjects, self).collect(objs, source_attr=source_attr, **kwargs)
+            return super().collect(objs, source_attr=source_attr, **kwargs)
         except models.ProtectedError as e:
             self.protected.update(e.protected_objects)
 
     def related_objects(self, related, objs):
-        qs = super(NestedObjects, self).related_objects(related, objs)
+        qs = super().related_objects(related, objs)
```
Comment (yakky) [inline filer/admin/patched/admin_utils.py:106]: This https://github.com/divio/django-filer/pull/1202/files#diff-71f149c65e62bb3b8fdcd50b4b00a845R106 is needed here

**Comment by FinalAngel** [comment]
@yakky you also included these changes:

* https://github.com/divio/django-filer/pull/1202/files#diff-71f149c65e62bb3b8fdcd50b4b00a845
* https://github.com/divio/django-filer/pull/1202/files#diff-d08b27fd1826e47143342eee8a59a02a
* https://github.com/divio/django-filer/pull/1202/files#diff-8d2f6449635dadfd7a030ac92dfb1fd5
* https://github.com/divio/django-filer/pull/1202/files#diff-2eeaed663bd0d25b7e608891384b7298R10
* https://github.com/divio/django-filer/pull/1202/files#diff-5b96e697e7a218b995b7fdf9b754c366

shall I port them too?

File: filer/migrations/0006_auto_20160623_1627.py [hunk 3]
```diff
@@ -1,7 +1,4 @@
-# -*- coding: utf-8 -*-
 # Generated by Django 1.9.7 on 2016-06-23 20:27
-from __future__ import unicode_literals
-
 import django.db.models.deletion
```
Comment (yakky) [inline filer/migrations/0006_auto_20160623_1627.py:2]: Change needed for 3.1 https://github.com/divio/django-filer/pull/1202/files#diff-8d2f6449635dadfd7a030ac92dfb1fd5R19

File: tests/utils/extended_app/migrations/0001_initial.py [hunk 4]
```diff
@@ -1,6 +1,4 @@
-# -*- coding: utf-8 -*-
 # Generated by Django 1.9.5 on 2016-04-20 13:41
```
Comment (yakky) [inline tests/utils/extended_app/migrations/0001_initial.py:1]: we'd better drop migrations for this app and create a new one to sync auto fields changes in 3.1

**Comment by yakky** [comment]
> @yakky you also included these changes:
> 
>     * https://github.com/divio/django-filer/pull/1202/files#diff-71f149c65e62bb3b8fdcd50b4b00a845
> 
>     * https://github.com/divio/django-filer/pull/1202/files#diff-d08b27fd1826e47143342eee8a59a02a
> 
>     * https://github.com/divio/django-filer/pull/1202/files#diff-8d2f6449635dadfd7a030ac92dfb1fd5
> 
>     * https://github.com/divio/django-filer/pull/1202/files#diff-2eeaed663bd0d25b7e608891384b7298R10
> 
>     * https://github.com/divio/django-filer/pull/1202/files#diff-5b96e697e7a218b995b7fdf9b754c366
> 
> 
> shall I port them too?

Yes, I pinpointed them in the review :)

**Comment by yakky** [comment]
we have to wait the release of this https://github.com/django-polymorphic/django-polymorphic/pull/452 for properly support 3.1, unfortunately

**Comment by FinalAngel** [comment]
awesome @yakky, I'll have a closer look tomorrow. Also cms will be dropped from the tests, that was mostly copy & paste. Not yet ready :) marked WIP.

And yeah, I added the first comment on polymorphic xD

**Comment by FinalAngel** [comment]
@yakky I'll remove the integration tests for now as they are not stable anymore. Needs more work in the future...
Would you know what's up with the documentation, I tried several code changes without any positive result...

**Comment by yakky** [comment]
> @yakky I'll remove the integration tests for now as they are not stable anymore. Needs more work in the future...
> Would you know what's up with the documentation, I tried several code changes without any positive result...

I see fe tests are enabled again now

I will work with 3baal to have a new release of polymorphic soon

I see docs tests failing due to missing sphinx

**Comment by FinalAngel** [comment]
@yakky sorry it's this one: https://travis-ci.org/github/divio/django-filer/jobs/718533457
I don't get why filer is not installed

File: filer/migrations/0001_initial.py [hunk 5]
```diff
@@ -148,7 +145,7 @@ class Migration(migrations.Migration):
         migrations.CreateModel(
             name='Image',
             fields=[
-                ('file_ptr', models.OneToOneField(serialize=False, auto_created=True, to='filer.File', primary_key=True, parent_link=True, on_delete=django.db.models.deletion.CASCADE)),
+                ('file_ptr', models.OneToOneField(serialize=False, auto_created=False, to='filer.File', primary_key=True, parent_link=True, on_delete=django.db.models.deletion.CASCADE)),
```
Comment (FinalAngel) [inline filer/migrations/0001_initial.py:148]: @yakky is that change required in all other addons as well? Like when it's a "OneToOneField" change auto_created to False?

File: filer/migrations/0006_auto_20160623_1627.py [hunk 6]
```diff
@@ -16,6 +13,6 @@ class Migration(migrations.Migration):
         migrations.AlterField(
             model_name='image',
             name='file_ptr',
-            field=models.OneToOneField(on_delete=django.db.models.deletion.CASCADE, primary_key=True, related_name='%(app_label)s_%(class)s_file', serialize=False, to='filer.File'),
+            field=models.OneToOneField(on_delete=django.db.models.deletion.CASCADE, primary_key=True, related_name='%(app_label)s_%(class)s_file', serialize=False, to='filer.File', parent_link=True),
```
Comment (FinalAngel) [inline filer/migrations/0006_auto_20160623_1627.py:16]: @yakky same here, do we need to apply it to other migrations as well?

File: filer/fields/folder.py [hunk 7]
```diff
@@ -100,6 +97,7 @@ def __init__(self, rel, queryset, to_field_name, *args, **kwargs):
         self.max_value = None
         self.min_value = None
         kwargs.pop('widget', None)
+        kwargs.pop('blank', None)
```
Comment (FinalAngel) [inline filer/fields/folder.py:100]: @yakky I'm not sure if this is the right approach, in 3.1 we get the following error otherwise while using djangocms-file:
  File "/Users/finalangel/Sites/_testing/djangocms380/mysite/packages/djangocms-file/djangocms_file/cms_plugins.py", line 36, in <module>
    class FolderPlugin(CMSPluginBase):
  File "/Users/finalangel/Sites/_testing/djangocms380/mysite/packages/django-cms/cms/plugin_base.py", line 61, in __new__
    new_plugin.form = type('%sForm' % name, (forms.ModelForm,), form_attrs)
  File "/Users/finalangel/.pyenv/versions/django-cms-3.8.0/lib/python3.6/site-packages/django/forms/models.py", line 258, in __new__
    apply_limit_choices_to=False,
  File "/Users/finalangel/.pyenv/versions/django-cms-3.8.0/lib/python3.6/site-packages/django/forms/models.py", line 175, in fields_for_model
    formfield = f.formfield(**kwargs)
  File "/Users/finalangel/Sites/_testing/djangocms380/mysite/packages/django-filer/filer/fields/folder.py", line 135, in formfield
    return super().formfield(**defaults)
  File "/Users/finalangel/.pyenv/versions/django-cms-3.8.0/lib/python3.6/site-packages/django/db/models/fields/related.py", line 994, in formfield
    'blank': self.blank,
  File "/Users/finalangel/.pyenv/versions/django-cms-3.8.0/lib/python3.6/site-packages/django/db/models/fields/related.py", line 419, in formfield
    return super().formfield(**defaults)
  File "/Users/finalangel/.pyenv/versions/django-cms-3.8.0/lib/python3.6/site-packages/django/db/models/fields/__init__.py", line 929, in formfield
    return form_class(**defaults)
  File "/Users/finalangel/Sites/_testing/djangocms380/mysite/packages/django-filer/filer/fields/folder.py", line 100, in __init__
    forms.Field.__init__(self, widget=self.widget(rel, site), *args, **kwargs)
TypeError: __init__() got an unexpected keyword argument 'blank'
```

```

File: filer/migrations/0001_initial.py (hunk 5 above)
Comment (yakky) [reply filer/migrations/0001_initial.py:148]: Sorry, the change from `auto_created=True` to `auto_created=False` is not needed (it's been left by error in my commits)
, only `parent_link=True` where missing is required

File: filer/migrations/0006_auto_20160623_1627.py (hunk 6 above)
Comment (yakky) [reply filer/migrations/0006_auto_20160623_1627.py:16]: wherever `file_ptr` is present in the migration we need `parent_link=True`

See also `tests/utils/custom_image/migrations/0002_auto_20160621_1510.py`

**Comment by yakky** [comment]
@FinalAngel I run local tests using soon-to-be released new version of polymorphic `https://github.com/django-polymorphic/django-polymorphic/archive/1d719a0e605632bb761fb002673c5530896e6f32.zip` and tests pass :top:

File: filer/fields/folder.py (hunk 7 above)
Comment (yakky) [reply filer/fields/folder.py:100]: *forms* `Field` does not have the `blank` attribute, I don't get why it's passed to `AdminFolderFormField`

Comment (FinalAngel) [reply filer/fields/folder.py:100]: @yakky I'm also getting that issue on other addons. It seems like something changed from Django<3.1 to 3.1. I need to add that also on the CMS. Let me know if you figure out more eventually.

**Comment by FinalAngel** [comment]
@yakky will wait for the release though 👍  Any ideas about the docs tests?

File: docs/conf.py [hunk 8]
```diff
@@ -11,13 +10,17 @@
...
-import sys, os
-sys.path.append(os.path.abspath('../'))
 import datetime
+import os
+import sys
 
 from filer import __version__
 
 
+sys.path.append(os.path.abspath('../'))
```
Comment (yakky) [inline docs/conf.py:20]: this must go above line 17

**Comment by yakky** [comment]
> @yakky will wait for the release though +1 Any ideas about the docs tests?

wrt docs tests https://github.com/divio/django-filer/pull/1201/files#r472112884

File: filer/fields/folder.py (hunk 7 above)
Comment (yakky) [reply filer/fields/folder.py:100]: I hope to be able to check this in a couple of days

as a stop gap solution, we can keep the current implementation, but I agree that it might there be more behind it

**Comment by FinalAngel** [comment]
@yakky this is ready from my end, would you be able to review and approve?

**Comment by FinalAngel** [comment]
awesome, once CMS passes we are ready for the release marathon :D

**Review by yakky — APPROVED** [review]
